pkg/nmstate/policy_test.go

```diff
@@ -0,0 +1,483 @@
+			expectedError:     "",
+			sriovInterface:    "ens1",
+			vlanID:            10,
+		},
+		{
+			testNMStatePolicy: buildValidPolicyTestBuilder(buildTestClientWithDummyPolicyObject()),
+			expectedError:     "nodenetworkconfigurationpolicy 'baseInterface' cannot be empty",
+			sriovInterface:    "",
+			vlanID:            10,
+		},
+		{
+			testNMStatePolicy: buildValidPolicyTestBuilder(buildTestClientWithDummyPolicyObject()),
+			expectedError:     "invalid vlanID, allowed vlanID values are between 0-4094",
+			sriovInterface:    "ens1",
+			vlanID:            4099,
+		},
+	}
+	for _, testCase := range testCases {
+		testPolicy := testCase.testNMStatePolicy.WithVlanInterface(testCase.sriovInterface, testCase.vlanID)
+		assert.Equal(t, testPolicy.errorMsg, testCase.expectedError)
+
+		desireState := &DesiredState{}
+		if testCase.expectedError == "" {
+			_ = yaml.Unmarshal(testPolicy.Definition.Spec.DesiredState.Raw, desireState)
+			assert.Equal(t, desireState, &DesiredState{
+				Interfaces: []NetworkInterface{
+					{
+						Name:  fmt.Sprintf("%s.%d", testCase.sriovInterface, testCase.vlanID),
+						Type:  "vlan",
+						State: "up",
+						Vlan: Vlan{
+							BaseIface: testCase.sriovInterface,
+							ID:        int(testCase.vlanID),
+						},
+					},
+				},
+			})
+		}
+	}
+}
+
+func TestPolicyWithAbsentInterface(t *testing.T) {
+	testCases := []struct {
+		testNMStatePolicy *PolicyBuilder
+		expectedError     string
+		baseInterface     string
+	}{
+		{
+			testNMStatePolicy: buildValidPolicyTestBuilder(buildTestClientWithDummyPolicyObject()),
+			expectedError:     "",
+			baseInterface:     "ens1",
+		},
+		{
+			testNMStatePolicy: buildValidPolicyTestBuilder(buildTestClientWithDummyPolicyObject()),
+			expectedError:     "nodenetworkconfigurationpolicy 'interfaceName' cannot be empty",
+			baseInterface:     "",
+		},
+	}
+	for _, testCase := range testCases {
+		testPolicy := testCase.testNMStatePolicy.WithAbsentInterface(testCase.baseInterface)
+		assert.Equal(t, testPolicy.errorMsg, testCase.expectedError)
+
+		desireState := &DesiredState{}
+		if testCase.expectedError == "" {
+			_ = yaml.Unmarshal(testPolicy.Definition.Spec.DesiredState.Raw, desireState)
+			assert.Equal(t, desireState, &DesiredState{
+				Interfaces: []NetworkInterface{{Name: testCase.baseInterface, State: "absent"}}})
+		}
+	}
+}
+
+func TestPolicyWithWithOptions(t *testing.T) {
+	testSettings := buildTestClientWithDummyPolicyObject()
+	testBuilder := buildValidPolicyTestBuilder(testSettings).WithOptions(
+		func(builder *PolicyBuilder) (*PolicyBuilder, error) {
+			return builder, nil
+		})
+
+	assert.Equal(t, "", testBuilder.errorMsg)
+	testBuilder = buildValidPolicyTestBuilder(testSettings).WithOptions(
+		func(builder *PolicyBuilder) (*PolicyBuilder, error) {
+			return builder, fmt.Errorf("error")
+		})
+
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func TestPolicyWaitUntilCondition(t *testing.T) {
+	testCases := []struct {
```

Comment:
I think there should also be a test case for when the policy does not exist
```go
{
	testNMStatePolicy: buildValidPolicyTestBuilder(clients.GetTestClients(clients.TestClientParams{})),
	expectedError:     fmt.Errorf("cannot wait for NodeNetworkConfigurationPolicy condition because it does not exist"),
	condition:         shared.NodeNetworkStateConditionAvailable,
},
```